### Inputs
The primary non-responsive data inputs to the Wild Worm Codon Adapter App are .csv files containing the following information:  
  1. Codon frequency rates and relative adaptiveness values for *S. ratti*, *C. elegans*, *N. brasiliensis*, and *B. malayi*  
  2. Optimal codon lookup table for *Strongyloides spp*, *Pristionchus spp* , *N. brasiliensis* , *B. malayi*, and *C. elegans* 
  3. <OPTIONAL> Custom optimal codon lookup table (2 columns: single-letter amino acid symbols and corresponding 3-letter optimal codon sequences; one optimal codon per amino acid)

Both .csv files described above are available for download.

### Codon Usage Rules
Codon bias in nematode transcripts can vary as a function of gene expression levels such that highly expressed genes appear to have the greatest degree of codon bias. Thus, codon frequency rates from highly expressed genes are used, whenever possible. Codon frequency rates for *Strongyloides* species are based on highly expressed *S. ratti* transcripts (50 most abundant expressed sequence tag clusters). Specifically, codon usage rules were generated by calculating the frequency for each codon from count data published in [Mitreva *et al* 2006](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1779591/); frequency values were manually checked against published frequency values. Codon frequency rates for *C. elegans* were based on highly expressed *C. elegans* gene count data published in [Sharp and Bradnam, 1997](https://www.ncbi.nlm.nih.gov/books/NBK20194/). Codon frequency rates for *B. malayi* and *N. brasiliensis* are based on count data from [Nematode.net](http://www.nematode.net/NN3_frontpage.cgi?navbar_selection=nemagene&subnav_selection=codon_usage_tables).  

### Relative Adaptiveness, Optimal Codons, and Optimization
For *S. ratti*, *C. elegans*, *N. brasiliensis*, and *B. malayi*: The relative adaptiveness values of every possible codon was generated as follows: individual codons were scored by calculating their relative adaptivness: (the frequency that codon "i" encodes amino acid "AA") / (the frequency of the codon most often used for encoding amino acid "AA"). Optimal codons for these species were defined as the codon with the highest relative adaptiveness value for each amino acid.

For *Pristionchus* species: Optimal codons were defined by codon usage bias calculations based on the top 10% highly expressed *P. pacificus* genes [(Han *et al*, 2020)](https://www.genetics.org/content/216/4/947).

User-provided custom optimization rules: In addition to the optimization rules provided by the application, users may also provide a custom set of optimal codons. In this case, users may upload a .csv file containing 2 columns listing single-letter amino acid symbols and the corresponding 3-letter optimal codon sequence, using the provided UI interface. Only one optimal codon should be provided per amino acid; stop codons should be designated using the '*' symbol. This custom optimal codon lookup table will be applied during codon optimization; CAI values will not be calculated. A template .csv file is available for download.  

In all cases, codon optimzation is performed by replacing non-optimal codons with optimal codons. 

### Codon Adaptation Index Values
In the case of optimization for *Strongyloides* species, *N. brasiliensis*, or *B. malayi*: sequences (both original and optimized) are scored by calculating the Codon Adaptation Index: the geometric average of relative adaptiveness of all codons in the gene sequence ( [Sharp and Li, 1987](https://pubmed.ncbi.nlm.nih.gov/3547335/), [Jansen *et al* 2003](http://www.ncbi.nlm.nih.gov/pubmed/12682375)). The CAI is calculated via the `seqinr` library using a multi-species relative adaptiveness table (see above). CAI values are not calculated when optimizating for *Pristionchus* species or when using a user-provided custom optimization rule.