### Inputs
The primary non-responsive data input to the *Strongyloides* Codon Adapter App is a .csv file containing codon usage rules for highly expressed *S. ratti* transcripts (50 most abundant expressed sequence tag clusters) and highly expressed *C. elegans* genes. The *C. elegans* codon usage rules were generated by calculating the frequency for each codon from count data published in [Sharp and Bradnam, 1997](https://www.ncbi.nlm.nih.gov/books/NBK20194/). For *S. ratti*, codon usage rules were generated by calculating the frequency for each codon from count data published in [Mitreva *et al* 2006](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1779591/); frequency values were manually checked against published frequency values. These calculations were done using a custom R script (`calculate_codon_usage_rules.R`) that has been uploaded to this projects GitHub repository (see: `Data` subfolder). For both species, .csv files containing count data are available for download, as is the multi-species codon usage chart that contains the calculated frequency values.

This multi-species codon usage chart is loaded by the Shiny server function and used to create relative adaptiveness lookup tables.  

### Codon Adaptation Index Values
For each sequence provided using responsive Shiny inputs, individual codons are scored by calculating their relative adaptivness: (the frequency that codon "i" encodes amino acid "AA") / (the frequency of the codon most often used for encoding amino acid "AA"). Genes are scored by calculating their Codon Adaptation Index: the geometric average of relative adaptiveness of all codons in the gene sequence ( [Sharp and Li, 1987](https://pubmed.ncbi.nlm.nih.gov/3547335/), [Jansen *et al* 2003](http://www.ncbi.nlm.nih.gov/pubmed/12682375)). The CAI is calculated via the `seqinr` library. Codon bias in nematode transcripts can vary as a function of gene expression  levels such that highly expressed genes appear to have the greatest degree of codon bias. Therefore, optimization rules used to generate sequences codon optimized for expression in *Strongyloides* species are based on the codon usage weights of highly expressed *S. ratti* transcripts [(Mitreva *et al* 2006)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1779591/).