### Inputs
The primary non-responsive data inputs to the Wild Worm Codon Adapter App are .csv files containing the following information:  
  1. Codon frequency rates and relative adaptiveness values for *S. ratti*, *N. brasiliensis*, *B. malayi*, *P. pacificus*, and *C. elegans* 
  2. Optimal codon lookup table for *Strongyloides spp*, *N. brasiliensis*, *B. malayi*, *P. pacificus*, and *C. elegans* 
  3. <OPTIONAL> Custom optimal codon lookup table (2 columns: single-letter amino acid symbols and corresponding 3-letter optimal codon sequences; one optimal codon per amino acid) 
  4. <OPTIONAL> Custom intron list (fasta file containing a maximum of 3 introns; intron sequences should begin/end with canonical 5'-GT<intron>AG-3' splice recognition sequences)

All files described above are available for download via the Data Availability panel.

### Codon Usage Rules
The codon usage patterns of highly expressed genes are thought to correlate with higher protein expression ( [Sharp and Li, 1987](https://pubmed.ncbi.nlm.nih.gov/3547335/) , [Plotkin and Kudla, 2011](http://www.nature.com/articles/nrg2899) ). Thus, codon frequency rates from highly expressed genes are used. 

Codon frequency rates for *Strongyloides* species are based on highly expressed *S. ratti* transcripts (50 most abundant expressed sequence tag clusters). Specifically, codon usage rules were generated by calculating the frequency for each codon from count data published in [Mitreva *et al* (2006)](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC1779591/); frequency values were manually checked against published frequency values. 

Codon frequency rates for *C. elegans* were based on highly expressed *C. elegans* gene count data published in [Sharp and Bradnam (1997)](https://www.ncbi.nlm.nih.gov/books/NBK20194/). 

Codon frequency rates for *N. brasiliensis* were calculated from coding sequences of highly expressed *N. brasiliensis* genes (10% highest RNA-seq expression values across all samples); RNA-seq data was downloaded from WormBase ParaSite, based on data originally published in [Eccles *et al* (2018)](https://bmcbiol.biomedcentral.com/articles/10.1186/s12915-017-0473-4). 

Codon frequency rates for highly expressed *B. malayi* and *P. pacificus* genes (average frequency in bins 8-11, ~10% highest expressing genes) are from [Han *et al* (2020)](https://www.genetics.org/content/216/4/947); raw codon frequency data were graciously provided by Dr. Wen-Sui Lo and Dr. Ralf Sommer.

### Relative Adaptiveness, Optimal Codons, and Optimization
The relative adaptiveness values for every possible codon were generated as follows. Individual codons were scored by calculating their relative adaptivness: (the frequency that codon "i" encodes amino acid "AA") / (the frequency of the codon most often used for encoding amino acid "AA"). Optimal codons for these species were defined as the codon with the highest relative adaptiveness value for each amino acid.

User-provided custom optimization rules: In addition to the optimization rules provided by the application, users may also provide a custom set of optimal codons. In this case, users may upload a .csv file containing 2 columns listing single-letter amino acid symbols and the corresponding 3-letter optimal codon sequence, using the provided UI interface. Only one optimal codon should be provided per amino acid; stop codons should be designated using the '*' symbol. This custom optimal codon lookup table will be applied during codon optimization; CAI values will not be calculated. A template .csv file is available for download.  

In all cases, codon optimization is performed by replacing non-optimal codons with optimal codons. 

### Codon Adaptation Index Values
Sequences (both original and optimized) are scored by calculating the Codon Adaptation Index: the geometric average of relative adaptiveness of all codons in the gene sequence ( [Sharp and Li, 1987](https://pubmed.ncbi.nlm.nih.gov/3547335/), [Jansen *et al* 2003](http://www.ncbi.nlm.nih.gov/pubmed/12682375)). The CAI is calculated via the `seqinr` library using a multi-species relative adaptiveness table (see above).
