---
title: "Codon Usage Chemoreceptor Adaptiveness"
date: "10/7/2020"
output:
  pdf_document:
    toc: true
    toc_depth: 3
    number_sections: true
  
---

# Introduction  
The purpose of this document is to compare codon usage for user-defined genes-of-interest against
the codon usage of every CDS in the genomes of four *Strongyloides spp.* (and possibly *C. elegans*).  

# Data Preprocessing  
Full lists of CDS sequences for *S. stercoralis*, *S. ratti*, *S. papillosus*, *S. venezuelensis* and *C. elegans* were download from Wormbase ParaSite on October 1, 2020. The .fa files were used as input to the *Strongyloides* Codon Adapter App. For each species an excel report containing the computed codon adaptation index values relative to *S. ratti* and *C. elegans* was downloaded; these file are uploaded the code chunks below.  

For user-defined genes-of-interest, a list of *S. stercoralis* chemoreceptor genes was used as input to the *Strongyloides* Codon Adapter App; the generated excel report is uploaded below.  

```{r setup, echo=FALSE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(tidyverse)
  library(readxl)
  library(magrittr)
  library(DescTools)
  library(ggthemes)
  library(biomaRt)
  library(ggplot2)
  library(openxlsx)
  library(ggplot2)
  library(wrapr)
  library(Matching)
  library(cowplot)
  source('KSUtils.R')
  source('PTUtils.R')
})
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

## Parse Species-specific lists of chemoreceptors 
Cleaning and data wrangling for chemoreceptor gene lists; necessary prepreocessing step before running lists of genes through the *Strongyloides* Codon Adapter App. Gene lists were downloaded from a Shiny app provided for this purpose by [Wheeler *et al* 2020](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3000723). Species for which we have lists of chemoreceptors are: *C. elegans*, *S. stercoralis*, *S. ratti*, and *S. venezuelensis*. Data are saved as 2-column .csv files containing geneIDs and cDNA sequences; these files can be used as inputs to the *Strongyloides* Codon Adapter App.     

```{r cleanGeneLists, eval = F}
# C. elegans ----
temp <- c(Ce = './Data/Chemoreceptors/Ce_chemoreceptors.csv')
genelist.Ce <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For C. elegans, match wormbase gene sequence name to the wbps geneID, 
## wbps transcriptID, 
## and pull the cDNA sequence
Ce.tr.seq <- fetch_cDNAseq(species = 'caelegprjna13758',
                           inputIDs = genelist.Ce$Transcript_ID,
                           inputFilter = 'wormbase_gseqname')

write.table(Ce.tr.seq, 
            file = "./Data/Chemoreceptors/Ce_chemosensory_cDNA_list.csv", 
            sep = ",", 
            col.names = FALSE, 
            row.names = FALSE)

# S. stercoralis ----
temp <- c(Ss = './Data/Chemoreceptors/Ss_chemoreceptors.csv')

genelist.Ss <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For stercoralis, match wormbase gene sequence name to the wbps geneID, 
## wbps transcriptID 
## and pull the cDNA sequence
Ss.tr.seq <- fetch_cDNAseq(species = 'ststerprjeb528',
                           inputIDs = genelist.Ss$Transcript_ID,
                           inputFilter = 'wbps_transcript_id')

write.table(Ss.tr.seq, 
            file = "./Data/Chemoreceptors/Ss_chemosensory_cDNA_list.csv", 
            sep = ",", 
            col.names = FALSE, 
            row.names = FALSE)


# S. ratti ----
temp <- c(Sr = './Data/Chemoreceptors/Sr_chemoreceptors.csv')

genelist.Sr <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For ratti, match wormbase gene sequence name to the wbps geneID, 
## wbps transcriptID 
## and pull the cDNA sequence
Sr.tr.seq <- fetch_cDNAseq(species = 'strattprjeb125',
                           inputIDs = genelist.Sr$Transcript_ID,
                           inputFilter = 'wbps_transcript_id')

write.table(Sr.tr.seq, 
            file = "./Data/Chemoreceptors/Sr_chemosensory_cDNA_list.csv", 
            sep = ",", 
            col.names = FALSE, 
            row.names = FALSE)

# S. venezuelensis ----
temp <- c(Sv = './Data/Chemoreceptors/Sv_chemoreceptors.csv')

genelist.Sv <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For venezuelnsis, match wormbase gene sequence name to the 
## wbps geneID, wbps transcriptID,
##  and pull the cDNA sequence
Sv.tr.seq <- fetch_cDNAseq(species = 'stveneprjeb530',
                           inputIDs = genelist.Sv$Transcript_ID,
                           inputFilter = 'wbps_transcript_id')

write.table(Sv.tr.seq, 
            file = "./Data/Chemoreceptors/Sv_chemosensory_cDNA_list.csv", 
            sep = ",", 
            col.names = FALSE, 
            row.names = FALSE)
```


# Analysis/Results   

## Filter Genome Data  
For each species, load data listing GC content and CAI indeces for the full list of CDS elements in genome. The full list contains transcript level information. In cases where there are multiple transcripts per gene, only use CAI values from the longest transcript. Save a new copy of the data. If the filtered versions are avaliable as files, don't regenerate them. 
```{r filterGenomeData, message=FALSE, include=FALSE}
if (!all(file.exists(c("./Data/Genomic/Ss_Genomic_CAI_Table.xlsx",
                       "./Data/Genomic/Sr_Genomic_CAI_Table.xlsx",
                       "./Data/Genomic/Sp_Genomic_CAI_Table.xlsx",
                       "./Data/Genomic/Sv_Genomic_CAI_Table.xlsx",
                       "./Data/Genomic/Ce_Genomic_CAI_Table.xlsx")))) {
  
  temp <- c(Ss = './Data/Genomic/Ss_Codon_Usage_Report.xlsx',
            Sr = './Data/Genomic/Sr_Codon_Usage_Report.xlsx',
            Sp = './Data/Genomic/Sp_Codon_Usage_Report.xlsx',
            Sv = './Data/Genomic/Sv_Codon_Usage_Report.xlsx',
            Ce = './Data/Genomic/Ce_Codon_Usage_Report.xlsx')
  dat.genomic <- sapply(temp, read_excel, 
                        sheet = 1, 
                        skip = 3, 
                        col_names = T, 
                        simplify = F, 
                        USE.NAMES = T)
  # This data contains transcript level information. In cases where there are 
  # multiple transcripts per gene, need to only include 
  # values from the longest transcript.
  dat.genomic.cleaned <- lapply(dat.genomic, function(x){
    x %>%
      dplyr::rename(transcriptID = geneID) %>%
      dplyr::mutate(geneID = str_remove_all(transcriptID, "\\.[0-9]$")) %>%
      dplyr::mutate(geneID = str_remove_all(geneID, "[a-c]$")) %>%
      dplyr::group_by(geneID) %>%
      dplyr::slice_max(n = 1, order_by = str_length(`cDNA sequence`), 
                       with_ties = FALSE) %>%
      dplyr::relocate(geneID, .before = everything())
    
  })
  
  # Save cleaned versions of the genomic datasets
  suppressMessages(lapply(names(dat.genomic.cleaned), function(x){
    y <- dat.genomic.cleaned[[x]]
    write.xlsx(y, 
               file = paste0("./Data/Genomic/", x,"_Genomic_CAI_Table.xlsx")
    )
    
  }))
  
}
```

## Tidy Genome Data 
Tidy the genomic CAI values for downstream plotting and analysis. This can be computationally intensive, so save a tidy version as a .csv file. If that file exists, don't re-run this code.  
```{r loadGenomeData, message=FALSE, include=FALSE}
if (!file.exists(c("./Data/Genomic/tidy_genomic_CAI.csv"))) {
  temp <- c("./Data/Genomic/Ss_Genomic_CAI_Table.xlsx",
            "./Data/Genomic/Sr_Genomic_CAI_Table.xlsx",
            "./Data/Genomic/Sp_Genomic_CAI_Table.xlsx",
            "./Data/Genomic/Sv_Genomic_CAI_Table.xlsx",
            "./Data/Genomic/Ce_Genomic_CAI_Table.xlsx")
  dat.genomic.cleaned <- sapply(temp, read_excel, 
                                sheet = 1, 
                                col_names = T, 
                                simplify = F, 
                                USE.NAMES = T)
  names(dat.genomic.cleaned) <- c("Ss", "Sr", "Sp", "Sv", "Ce")
  
  dat.genomic.df <- dat.genomic.cleaned %>%
    map("geneID") %>%
    unlist() %>%
    as_tibble_col(column_name = "geneID") %>%
    group_by(geneID)
  
  dat.genomic.df <- dat.genomic.cleaned %>%
    map("transcriptID") %>%
    unlist() %>%
    as_tibble_col(column_name = "transcriptID") %>%
    add_column(dat.genomic.df, .)
  
  dat.genomic.df <- dat.genomic.cleaned %>%
    map("Sr_CAI") %>%
    unlist() %>%
    as_tibble_col(column_name = "Sr_CAI") %>%
    add_column(dat.genomic.df, .)
  
  dat.genomic.df <- dat.genomic.cleaned %>%
    map("Ce_CAI") %>%
    unlist() %>%
    as_tibble_col(column_name = "Ce_CAI") %>%
    add_column(dat.genomic.df, .)
  
  dat.genomic.df <- dat.genomic.cleaned %>%
    map("GC") %>%
    unlist() %>%
    as_tibble_col(column_name = "GC") %>%
    add_column(dat.genomic.df, .)
  
  # Add the column that species which species the gene is from
  # by matching to the original lists of genes, 
  # the group
  dat.genomic.df <- dat.genomic.df %>%
    dplyr::mutate(species = case_when(
      transcriptID %in% dat.genomic$Ss$geneID ~ 'Ss',
      transcriptID %in% dat.genomic$Sr$geneID ~ 'Sr',
      transcriptID %in% dat.genomic$Sp$geneID ~ 'Sp',
      transcriptID %in% dat.genomic$Sv$geneID ~ 'Sv',
      transcriptID %in% dat.genomic$Ce$geneID ~ 'Ce',
    ))
  dat.genomic.df <- group_by(dat.genomic.df, species)
  
  write.table(dat.genomic.df, 
              file = "./Data/Genomic/tidy_genomic_CAI.csv", 
              sep = ",", 
              col.names = T, 
              row.names = FALSE)
}
```

## Plotting Distribution of CAI Values  
This section takes the calculated codon adaptation index values for all predicted codoning sequences, and generates a density curve for each species. Ideally, this plot could be used to benchmark the calculated CAI values against other metrics of codon bias in these species. For example, Mitreva *et al* 2006 reported effective number of codons (ENC) as a measure of gene-wise codon bias, across species. They report that "mean ENC across all sampled nemtaode species is 46.7 +/- 5.1" but that *S. stercoralis* and *S. ratti* are outliers with low ENC values (~35). Note: ENC values range from 20 for highly biased to 61 for no bias. Thus, benchmarking from Mitreva *et al* 2006, we might expect that *Strongyloides* species will have greater codon bias than *C. elegans*. Unfortunately, although they do report measuring distribution of ENC values across all transcripts for each species, they only have plots for *S. ratti*, *P. trichosuri* and *P. pacificus*.   
```{r caiDist}
dat.genomic.df <- suppressWarnings(
  read.csv("./Data/Genomic/tidy_genomic_CAI.csv", 
           header = TRUE)) %>%
  as_tibble() 

species.specific.dat.df <- dat.genomic.df %>%
  dplyr::mutate(species_CAI = case_when(
    species == 'Ce' ~ Ce_CAI,
    species != 'Ce' ~ Sr_CAI
  )) %>%
  dplyr:: select(geneID, species_CAI, species) %>%
  group_by(species)

plot_distributions <- function(dat,
                               title,
                               xlabel,
                               ylabel = "Density (scaled to maximum of 1)",
                               xlim = c(0,1),
                               ylim = c(0,1)) {
  dat %>%
  ggplot(aes(x = species_CAI, y = after_stat(ndensity), color = species)) +
  geom_freqpoly(bins = 50) +
  scale_color_manual(values = c("seagreen4", "steelblue4", 
                                "coral4", "darkgoldenrod4", 
                                "darkorchid4"),
                     name = "Species",
                     labels = c("C. elegans", "S. papillosus", 
                                "S. ratti", "S. stercoralis",
                                "S. venezuelensis")) +
  labs(title = title,
       subtitle = "Data includes all predicted coding sequences",
       x = xlabel,
       y = ylabel,
       caption = "Strongyloides species values relative to S. ratti usage rules
       C. elegans values relative to C. elegans usage rules.") +
  coord_equal(xlim = xlim, ylim = ylim) +
  theme_bw() +
  theme(plot.title.position = "plot",
        plot.caption.position = "panel",
        plot.title = element_text(face = "bold",
                                  size = 10, hjust = 0),
        plot.subtitle = element_text(size = 8),
        legend.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 1))
}
species_adaptiveness_plot <- plot_distributions(species.specific.dat.df,
                                                "Species-wide codon adaptiveness",
                                                "Codon bias relative to \n genus coding usage rules (CAI)")
  

ggsave('./Plots/Codon_Adaptiveness_Distributions_by_Species.pdf', 
       plot = species_adaptiveness_plot, 
       width = 5.5, height = 4, 
       units = "in", device = "pdf", 
       useDingbats = FALSE)

species_adaptiveness_plot
```

Notes re: the above plot - *C. elegans* shows overall lower codon bias compared to the *Strongyloides* species, as predicted from the Mitreva *et al* 2006 results. Reminder that Mitreva *et al* reported that this difference is greatly affected by the GC content; more AT rich speices like the *Strongyloides* species have greater codon usage biases. Also, it looks like the *Strongyloides* subclades (*S. ratti* - *S. stercoralis*; *S. papillosus* - *S. venezuelensis*) cluster together.  

### Kruskal-Wallis Test with Dunn's Post tests for differences in genomic CAI values across species  
```{r genomic.CAI.kwtest}
# Kruskal-Wallis testing of genomic CAI
one.way <- kruskal.test(species_CAI ~ species, data = species.specific.dat.df)

# Posthoc Dunn tests with Bonferroni adjustment
post.hoc <- dunn.test::dunn.test(species.specific.dat.df$species_CAI, species.specific.dat.df$species, method = "bonferroni")
```

## Load Genes-of-Interest Data  
For gene set of interest, load data listing GC content and CAI indeces.  
```{r loadGoIData, message=FALSE}
temp <- c(
  Ss_GoI = './Data/Chemoreceptors/Ss_Chemoreceptor_Codon_Usage_Report.xlsx',
  Sr_GoI = './Data/Chemoreceptors/Sr_Chemoreceptor_Codon_Usage_Report.xlsx',
  Sv_GoI = './Data/Chemoreceptors/Sv_Chemoreceptor_Codon_Usage_Report.xlsx',
  Ce_GoI = './Data/Chemoreceptors/Ce_Chemoreceptor_Codon_Usage_Report.xlsx')

dat.Ss.GoI <- read_excel(temp[["Ss_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Ss")

dat.Sr.GoI <- read_excel(temp[["Sr_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Sr")

dat.Sv.GoI <- read_excel(temp[["Sv_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Sv")

dat.Ce.GoI <- read_excel(temp[["Ce_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Ce")

dat.GoI.df <- full_join(dat.Ss.GoI, dat.Sr.GoI, by = NULL) %>%
  full_join(dat.Sv.GoI, by = NULL) %>%
  full_join(dat.Ce.GoI, by = NULL) %>%
  group_by(species)

# Genome-wide CAI values = species.specific.dat.df

species.specific.GoI.dat.df <- dat.GoI.df %>%
  dplyr::mutate(species_CAI = case_when(
    species == 'Ce' ~ Ce_CAI,
    species != 'Ce' ~ Sr_CAI
  )) %>%
  dplyr:: select(geneID, species_CAI, species) %>%
  group_by(species)

```

## Plot Genes-of-Interest CAI values relative to Genomic data  
Generate a series of plots (separated by species ID) where for each gene the codon adaptation index relative to highly expressed *S. ratti* transcripts is plotted against the codon adaptation index relative to highly expressed *C. elegans* genes. For every plot, also include the values for the genes-of-interest set, so we can compare how the values of those genes fall compared to all the other genes in that respective species' genome.  
```{r plotGoIData}

tbl <- dat.genomic.df

cai_plot <- ggplot(tbl, aes(Sr_CAI, Ce_CAI, species)) +
  geom_smooth(method=lm, color = "steelblue4", 
              fill = "steelblue1", linetype = 2, 
              size = 0.5, formula = "y ~ x") +
  geom_point(shape = 1, size = 2, alpha = 0.5) +
  geom_point(dat.GoI.df, mapping = aes(Sr_CAI, Ce_CAI, color = species),
             show.legend = F,
             shape = 1, size = 2, alpha = 1) +
  scale_color_manual(values = c("seagreen4", 
                                "coral4", 
                                "darkgoldenrod4", 
                                "darkorchid4"))+
  
  geom_hline(yintercept = 0.5, color = "grey", linetype = 3) +
  geom_vline(xintercept = 0.5, color = "grey", linetype = 3) +
  facet_grid(~species) +
  labs(title = "Species-specific codon adaptiveness",
       subtitle = paste("Color icons = species-specific chemoreceptors;",
                        "Black icons = all coding sequences
         "),
       x = "Codon bias relative to \n S. ratti usage rules (CAI)",
       y = "Codon Bias relative to \n C. elegans usage rules (CAI)",
       caption = "Blue line/shading = linear regression
                       w/ 95% confidence regions;
                       formula = y ~ x") +
  coord_equal(xlim = c(0,1), ylim = c(0,1)) +
  theme_bw() +
  theme(plot.title.position = "plot",
        plot.caption.position = "panel",
        plot.title = element_text(face = "bold",
                                  size = 10, hjust = 0),
        plot.subtitle = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 1))

ggsave('./Plots/Chemoreceptor_Codon_Usage_Scatter_Plot.pdf', plot = cai_plot, 
       width = 11, height = 8, 
       units = "in", device = "pdf", 
       useDingbats = FALSE)

cai_plot

```

<!-- ### Kruskal-Wallis Test with Dunn's Post tests for differences in Chemoreceptor CAI values across species   -->
<!-- ```{r GoI.CAI.kwtest} -->
<!-- # Kruskal-Wallis testing of chemoreceptor CAI -->
<!-- one.way.GoI <- kruskal.test(species_CAI ~ species, data = species.specific.GoI.dat.df) -->

<!-- # Posthoc Dunn tests with Bonferroni adjustment -->
<!-- post.hoc.GoI <- dunn.test::dunn.test(species.specific.GoI.dat.df$species_CAI, species.specific.GoI.dat.df$species, method = "bonferroni") -->
<!-- post.hoc.GoI -->
<!-- ``` -->

## Plotting Distribution of Chemoreceptor CAI Values versus Genome-wide CAI Values 
This section takes the calculated codon adaptation index values for chemoreceptor sequences and all predicted codoning sequences, and generates density curve for each data subset, for each species. The goal of this plot is to see whether there are systematic differences in the density of CAI values in the subset of chemoreceptor genes.   
```{r caiGoIDist}

plot_GvS_distributions <- function(genomic.dat, 
                                   species.dat, title, 
                                   xintercept = 0.5, xlim = c(0,1), 
                                   yintercept = 0.5, ylim = c(0,1)) {
  genomic.dat %>%
  ggplot(aes(x = species_CAI, y = after_stat(ndensity), color = species)) +
  geom_hline(yintercept = yintercept, color = "grey", linetype = 3) +
  geom_vline(xintercept = xintercept, color = "grey", linetype = 3) +
  geom_freqpoly(bins = 50, linetype = 2, show.legend = F) +
  geom_freqpoly(species.dat, 
                mapping = aes(x = species_CAI, 
                              y = after_stat(ndensity), 
                              color = species),
                bins = 50,  show.legend = F) +
  scale_color_manual(values = c("seagreen4", "steelblue4", 
                                "coral4", "darkgoldenrod4", 
                                "darkorchid4"),
                     name = "Species",
                     labels = c("C. elegans", "S. papillosus", 
                                "S. ratti", "S. stercoralis",
                                "S. venezuelensis")) +
  facet_grid(~species) +
  labs(title = title,
       subtitle = "Dashed lines = all genes; Solid lines = chemoreceptor genes
       ",
       x = "Codon bias relative to \n genus coding usage rules (CAI)",
       y = "Density \n (scaled to max of 1)",
       caption = "Strongyloides species values relative to S. ratti usage rules
       C. elegans values relative to C. elegans usage rules.") +
  coord_equal(xlim = xlim, ylim = ylim) +
  theme_bw() +
  theme(plot.title.position = "plot",
        plot.caption.position = "panel",
        plot.title = element_text(face = "bold",
                                  size = 10, hjust = 0),
        plot.subtitle = element_text(size = 8),
        legend.title = element_text(size = 8),
        axis.title = element_text(size = 8),
        axis.text = element_text(size = 8),
        axis.text.x = element_text(angle = 45, hjust = 1))
}
goi_species_adaptiveness_plot <- plot_GvS_distributions(
  species.specific.dat.df,
  species.specific.GoI.dat.df,
  "Comparing chemoreceptor CAI vs genome-wide CAI values")

ggsave('./Plots/Chemoreceptor_vs_Genomic_CAI_Distributions.pdf', 
       plot = goi_species_adaptiveness_plot, 
       width = 11, height = 8, 
       units = "in", device = "pdf", 
       useDingbats = FALSE)

goi_species_adaptiveness_plot
```

## Quantifying differences in CAI value distributions between whole genome and genes-of-interest  
### Two-Sample KS Testing and Mann-Whitney Tests  
I want to know whether the distrubtions plotted above are different. One possibility: test the two groups using of a two-sample Kolmogorov-Smirnov test or a Mann-Whitney Test. I initially set these up as simple tests of gene-of-interest samples vs genomic samples, but after consulting with Elissa we felt that this approach might not be as statistically robust as we might prefer (due in large part of the sensitivity of KS tests). I've preserved the original testing code, but see subsequent sections for versions that use permutation tests.  
```{r ksTesting}
# Genome-wide CAI values = species.specific.dat.df
# Genes-of-interest CAI values = species.specific.GoI.dat.df
GoI.dat.split<-group_split(species.specific.GoI.dat.df)
GoI.dat.split.keys <- group_keys(species.specific.GoI.dat.df)
names(GoI.dat.split) <-GoI.dat.split.keys$species

genome.dat.split<-group_split(species.specific.dat.df)
genome.dat.split.keys <- group_keys(species.specific.dat.df)
names(genome.dat.split) <-genome.dat.split.keys$species

# Kolmogorov-Smirnov test to see if the 
# two datasets are from the same underlying distribution
ks.result.list <- lapply(GoI.dat.split.keys$species, function(x){
  ks.test(genome.dat.split[[x]]$species_CAI, 
          GoI.dat.split[[x]]$species_CAI)
})
names(ks.result.list) <-GoI.dat.split.keys$species

# Mann-Whitney test to compare the discrepency in the median values
mw.result.list <- lapply(GoI.dat.split.keys$species, function(x){
  wilcox.test(genome.dat.split[[x]]$species_CAI,
              GoI.dat.split[[x]]$species_CAI)
})
names(mw.result.list) <-GoI.dat.split.keys$species

```

<!-- ### Kolmogorov-Smirnov Test Results   -->
<!-- This statistical test can detect if the two groups are drawn from different underlying distributions.  -->

<!-- #### *C. elegans*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(ks.result.list$Ce)  -->
<!-- ``` -->

<!-- #### *S. stercoralis*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(ks.result.list$Ss)  -->
<!-- ``` -->

<!-- #### *S. ratti*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(ks.result.list$Sr)  -->
<!-- ``` -->

<!-- #### *S. venezuelensis*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(ks.result.list$Sv)  -->
<!-- ``` -->

<!-- ### Mann-Whitney Test Results   -->
<!-- This statistical test can detect if the median values differ between the two groups.   -->

<!-- #### *C. elegans*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(mw.result.list$Ce)  -->
<!-- ``` -->

<!-- #### *S. stercoralis*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(mw.result.list$Ss)  -->
<!-- ``` -->

<!-- #### *S. ratti*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(mw.result.list$Sr)  -->
<!-- ``` -->

<!-- #### *S. venezuelensis*:   -->
<!-- ```{r message = TRUE, comment = ''}  -->
<!-- print(mw.result.list$Sv)  -->
<!-- ``` -->

### Permutation Testing using KS Statistic and Mean Differences
The above versions compared a small group (chemoreceptor CAI values) to a large set (genomic CAI values). This analysis is perhaps more robustly accomplished by asking: what is the likelihood that a random assignment of genomic/GoI identities recaptures the differences in the populations? We can do this first by testing differences in distriubtion with KS-statistics, and second by looking at differences between mean CAI values between the two groups.  

Put another way,  
For Large set X and subset Y:
Sample X uniformly B times, producing a subset Xb of X the size of Y  
Calcuate the test statisic Db from Xb and X.  
This gives an empircal distribution of the values Db.  
Calculate D* as the test statistic from Y and X.  
How many of Db are at least as far away from zero as D*.  
Approximate p-value: p = #{Db >= D*}/B  

Note: it appears that a similar method for permutation-based KS testing is implemented in the KS test supplied by GraphPad Prism.   

For all permutation tests below, the number of permutations = 10,000, and the p-value cutoff at 0.001. P-values above the cut-off will be catergorized as "not significantly different".  

#### Permutation Test of KS Statistic
```{r permutation.KS, cache = TRUE, dependson= c("ksTesting","loadGoIData","caiDist")}

# For Large set X and subset Y:
# Sample X uniforming 1000 times, producing a subset Xb of X the size of Y
# Calculate the KS statistic Db from Xb and X.
# This gives an empircal distribution of the values Db. 
# 
# Calculate D* as the KS statistic from Y and X
# How many of Db are larger than or equal to D*.
# Approximate p-value: p = #{Db >= D*}/B

# In other words, this is a permutation Test that asks: 
# do we see KS distances larger than D* often when 
# we draw two samples of the same size as my originals (X and Y) from the 
# same underlying distribution. 
nboots <- 10000
pthresh <- 0.001

ks.pt.result.list <- lapply(GoI.dat.split.keys$species, function(x){
  dcomp.ks <- boot_dist_compare(
    GoI.dat.split[[x]]$species_CAI, 
    genome.dat.split[[x]]$species_CAI,
    nboots)
  display_dist_compare(dcomp.ks, 
                       pthresh, 
                       paste(x, 
                             "- permutation test, GoI - genomic distribution"))
})
names(ks.pt.result.list) <-GoI.dat.split.keys$species

plot_grid(plotlist = ks.pt.result.list, 
          labels = c("A", "B", "C", "D"), 
          label_size = 12)

```

\newpage
#### Permutation Test of Difference in Mean CAI
```{r permutation.mean, cache = TRUE, dependson= c("ksTesting","loadGoIData","caiDist")}
mean.pt.result.list <- lapply(GoI.dat.split.keys$species, function(x){
  dcomp.mean <-  boot_dist_compare_mean(
    GoI.dat.split[[x]]$species_CAI,
    genome.dat.split[[x]]$species_CAI,
    nboots)
  display_dist_compare_mean(dcomp.mean, 
                            pthresh, 
                            paste(x, 
                                  "- permutation test, GoI - genomic mean CAI"))
})
names(mean.pt.result.list) <-GoI.dat.split.keys$species

plot_grid(plotlist = mean.pt.result.list, 
          labels = c("A", "B", "C", "D"), 
          label_size = 12)

```

\newpage
#### Permutation Test of Mean-centered KS Statistic 
Another option might be: if we center the values, we can use the K-S test to see if the chemoreceptor sample is just a shifted version of the distribution for the genomic values.       
```{r meancentered.KS, cache = TRUE, dependson= c("ksTesting","loadGoIData","caiDist")}

# mean-centering values
center_scale <- function(x){
  scale(x, scale = FALSE)
}
species.specific.dat.df.centered<-species.specific.dat.df %>%
  dplyr::mutate(species_CAI = center_scale(species_CAI))
species.specific.GoI.dat.df.centered <- species.specific.GoI.dat.df %>%
  dplyr::mutate(species_CAI = center_scale(species_CAI))

GoI.dat.split.centered<-group_split(species.specific.GoI.dat.df.centered)
names(GoI.dat.split.centered) <-GoI.dat.split.keys$species

genome.dat.split.centered<-group_split(species.specific.dat.df.centered)
names(genome.dat.split.centered) <-genome.dat.split.keys$species

goi_species_adaptiveness_plot_centered <- plot_GvS_distributions(
  species.specific.dat.df.centered,
  species.specific.GoI.dat.df.centered,
  "Comparing centered chemoreceptor CAI vs genome-wide CAI values", 
  xlim = c(-0.5, 0.5), xintercept = 0)
goi_species_adaptiveness_plot_centered 

# Kolmogorov-Smirnov test to see if the two datasets 
# are from the same underlying distribution
nboots <- 10000
pthresh <- 0.001

ks.mean.pt.result.list <- lapply(GoI.dat.split.keys$species, function(x){
  dcomp.ks <- boot_dist_compare(center_scale(GoI.dat.split[[x]]$species_CAI),
                                center_scale(genome.dat.split[[x]]$species_CAI),
                                nboots)
  display_dist_compare(dcomp.ks, 
                       pthresh, 
                       paste(x, 
              "- KS permut. test, GoI - genomic (mean-centered)"))
})
names(ks.mean.pt.result.list) <-GoI.dat.split.keys$species

plot_grid(plotlist = ks.mean.pt.result.list, 
          labels = c("A", "B", "C", "D"), 
          label_size = 12)
```
Taken together, the three permutation tests suggest that in the case of *C. elegans*, chemoreceptor genes show signifcantly lower mean CAI than the entire genome, and that in addition to a shift in the mean, the distribution of values is narrower (the later perhaps what one might expect in a situation where we are looking at closely related genes). For the *Strongyloides spp.*, we find that although the overall distributions are statistically different (see: permutation tests of the KS-statistics), once we test individual components, we find that there are not significant differences in mean values and that the mean-centered distributions are not significantly different. This suggests to me that the overall change noted by the full KS test is due to the combination of smaller shifts in mean and distribution.  

What then is the takeaway here? The most robust finding is likely that the relationship of the *C. elegans* chemoreceptors to the rest of the genome is distinct from that of *Strongyloides spp.*. How to statistically describe this relationship? Perhaps by computing for each species the difference between each chemoreceptor CAI value and the median genomic CAI value (producing a set of genomic-avg-centered chemoreceptor CAI values); then perform Krusktall_Wallis testing w/ Dunn's posttests between the species.     

### Comparing magnitude of discrepency between chemoreceptor/genomic CAI across species
```{r crossSpeciesCAI}
median.species.CAI <- dplyr::summarise(species.specific.dat.df, avg = median(species_CAI))
species.specific.GoI.avg_corrected.dat.df<-species.specific.GoI.dat.df %>%
  right_join(median.species.CAI, by = "species") %>%
  dplyr::mutate(species_CAI = species_CAI - avg) 

avg_corrected_adaptiveness_plot <- plot_distributions(species.specific.GoI.avg_corrected.dat.df,
                                                "Relative codon adaptiveness of chemoreceptor genes",
                                                xlabel = "Codon bias relative to \n genus coding usage rules (CAI) \n normalized to median genomic CAI",
                                                xlim = c(-0.5,0.5))

avg_corrected_adaptiveness_plot
# Kruskal-Wallist testing of genomic-avg centered chemoreceptor CAI
one.way.GoI <- kruskal.test(species_CAI ~ species, data = species.specific.GoI.avg_corrected.dat.df)

# Posthoc Dunn tests with Bonferroni adjustment
post.hoc.GoI <- dunn.test::dunn.test(species.specific.GoI.avg_corrected.dat.df$species_CAI,
                                     species.specific.GoI.avg_corrected.dat.df$species, 
                                     method = "bonferroni")

```

By this test, it appears that the offset of chemoreceptor genes from the species median CAI is significantly greater in *C. elegans* than in the *Strongyloides spp*.  

\newpage
# Appendix: All code for this report  
```{r ref.label=knitr::all_labels(), echo=TRUE, eval=FALSE}
```
