---
title: "Codon Usage Chemoreceptor Adaptiveness"
date: "10/2/2020"
output:
  html_document:
    code_folding: hide
    df_print: paged
    toc: yes
  html_notebook:
    toc: yes
---

# Introduction  
The purpose of this document is to compare codon usage for user-defined genes-of-interest against
the codon usage of every CDS in the genomes of four *Strongyloides spp.* (and possibly *C. elegans*).  

# Data Preprocessing  
Full lists of CDS sequences for *S. stercoralis*, *S. ratti*, *S. papillosus*, *S. venezuelensis* and *C. elegans* were download from Wormbase ParaSite on October 1, 2020. The .fa files were used as input to the *Strongyloides* Codon Adapter App. For each species an excel report containing the computed codon adaptation index values relative to *S. ratti* and *C. elegans* was downloaded; these file are uploaded the code chunks below.  

For user-defined genes-of-interest, a list of *S. stercoralis* chemoreceptor genes was used as input to the *Strongyloides* Codon Adapter App; the generated excel report is uploaded below.  

```{r setup, include=FALSE}
suppressPackageStartupMessages({
  library(knitr)
  library(rmarkdown)
  library(tidyverse)
  library(readxl)
  library(magrittr)
  library(DescTools)
  library(ggthemes)
  library(biomaRt)
  library(ggplot2)
})
knitr::opts_chunk$set(echo = TRUE)
```

## Parse Species-specific lists of chemoreceptors 
Cleaning and data wrangling for chemoreceptor gene lists; necessary prepreocessing step before running lists of genes through the *Strongyloides* Codon Adapter App. Gene lists were downloaded from a Shiny app provided for this purpose by [Wheeler *et al* 2020](https://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.3000723). Species for which we have lists of chemoreceptors are: *C. elegans*, *S. stercoralis*, *S. ratti*, and *S. venezuelensis*. Data are saved as 2-column .csv files containing geneIDs and cDNA sequences; these files can be used as inputs to the *Strongyloides* Codon Adapter App.     

```{r cleanGeneLists, eval = F}
# C. elegans ----
temp <- c(Ce = '../Data/Ce_chemoreceptors.csv')
genelist.Ce <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For C. elegans, match wormbase gene sequence name to the wbsp transcript id, and pull the cDNA sequence
Ce.tr.seq <- getBM(attributes=c('wbps_transcript_id', 'cdna'),
                   # grab the cDNA sequences for the given genes from WormBase Parasite
                   mart = useMart(biomart="parasite_mart",
                                  dataset = "wbps_gene",
                                  host="https://parasite.wormbase.org",
                                  port = 443),
                   filters = c('species_id_1010',
                               'wormbase_gseqname'),
                   values = list('caelegprjna13758',
                                 genelist.Ce$Transcript_ID),
                   useCache = F) %>%
  as_tibble() %>%
  #we need to rename the columns retreived from biomart
  dplyr::rename(geneID = wbps_transcript_id, cDNA = cdna)
Ce.tr.seq$cDNA <- tolower(Ce.tr.seq$cDNA)

write.table(Ce.tr.seq, file = "./Ce_chemosensory_cDNA_list.csv", sep = ",", col.names = FALSE, row.names = FALSE)

# S. ratti ----
temp <- c(Sr = '../Data/Sr_chemoreceptors.csv')

genelist.Sr <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For ratti, match wormbase gene sequence name to the wbsp transcript id, and pull the cDNA sequence
Sr.tr.seq <- getBM(attributes=c('wbps_transcript_id', 'cdna'),
                   # grab the cDNA sequences for the given genes from WormBase Parasite
                   mart = useMart(biomart="parasite_mart",
                                  dataset = "wbps_gene",
                                  host="https://parasite.wormbase.org",
                                  port = 443),
                   filters = c('species_id_1010',
                               'wbps_transcript_id'),
                   values = list('strattprjeb125',
                                 genelist.Sr$Transcript_ID),
                   useCache = T) %>%
  as_tibble() %>%
  #we need to rename the columns retreived from biomart
  dplyr::rename(geneID = wbps_transcript_id, cDNA = cdna)
Sr.tr.seq$cDNA <- tolower(Sr.tr.seq$cDNA)

write.table(Sr.tr.seq, file = "./Sr_chemosensory_cDNA_list.csv", sep = ",", col.names = FALSE, row.names = FALSE)

# S. venezuelensis ----
temp <- c(Sv = '../Data/Sv_chemoreceptors.csv')

genelist.Sv <- suppressWarnings(read.csv(temp, 
                                         header = TRUE, 
                                         colClasses = "character", 
                                         strip.white = T)) %>%
  as_tibble() 

## For ratti, match wormbase gene sequence name to the wbsp transcript id, and pull the cDNA sequence
Sv.tr.seq <- getBM(attributes=c('wbps_transcript_id', 'cdna'),
                   # grab the cDNA sequences for the given genes from WormBase Parasite
                   mart = useMart(biomart="parasite_mart",
                                  dataset = "wbps_gene",
                                  host="https://parasite.wormbase.org",
                                  port = 443),
                   filters = c('species_id_1010',
                               'wbps_transcript_id'),
                   values = list('stveneprjeb530',
                                 genelist.Sv$Transcript_ID),
                   useCache = T) %>%
  as_tibble() %>%
  #we need to rename the columns retreived from biomart
  dplyr::rename(geneID = wbps_transcript_id, cDNA = cdna)
Sv.tr.seq$cDNA <- tolower(Sv.tr.seq$cDNA)

write.table(Sv.tr.seq, file = "./Sv_chemosensory_cDNA_list.csv", sep = ",", col.names = FALSE, row.names = FALSE)
```


# Analysis/Results   

## Load Genome Data  
For each species, load data listing GC content and CAI indeces for the full list of CDS elements in genome.  
```{r loadGenomeData}
temp <- c(Ss = './Ss_Codon_Usage_Report.xlsx',
          Sr = './Sr_Codon_Usage_Report.xlsx',
          Sp = './Sp_Codon_Usage_Report.xlsx',
          Sv = './Sv_Codon_Usage_Report.xlsx',
          Ce = './Ce_Codon_Usage_Report.xlsx')
dat.genomic <- sapply(temp, read_excel, 
                      sheet = 1, 
                      skip = 3, 
                      col_names = T, 
                      simplify = F, 
                      USE.NAMES = T)

dat.genomic.df <- dat.genomic %>%
  map("geneID") %>%
  unlist() %>%
  as_tibble_col(column_name = "geneID") %>%
  group_by

dat.genomic.df <- dat.genomic %>%
  map("Sr_CAI") %>%
  unlist() %>%
  as_tibble_col(column_name = "Sr_CAI") %>%
  add_column(dat.genomic.df, .)

dat.genomic.df <- dat.genomic %>%
  map("Ce_CAI") %>%
  unlist() %>%
  as_tibble_col(column_name = "Ce_CAI") %>%
  add_column(dat.genomic.df, .)

dat.genomic.df <- dat.genomic %>%
  map("GC") %>%
  unlist() %>%
  as_tibble_col(column_name = "GC") %>%
  add_column(dat.genomic.df, .)

# Add the column that species which species the gene is from
# by matching to the original lists of genes, 
# the group
dat.genomic.df <- dat.genomic.df %>%
  dplyr::mutate(species = case_when(
    geneID %in% dat.genomic$Ss$geneID ~ 'Ss',
    geneID %in% dat.genomic$Sr$geneID ~ 'Sr',
    geneID %in% dat.genomic$Sp$geneID ~ 'Sp',
    geneID %in% dat.genomic$Sv$geneID ~ 'Sv',
    geneID %in% dat.genomic$Ce$geneID ~ 'Ce',
  ))
dat.genomic.df <- group_by(dat.genomic.df, species)


```

## Plotting Distribution of CAI Values  
This section takes the calculated codon adaptation index values for all predicted codoning sequences, and generates a density curve for each species. Ideally, this plot could be used to benchmark the calculated CAI values against other metrics of codon bias in these species. For example, Mitreva *et al* 2006 reported effective number of codons (ENC) as a measure of gene-wise codon bias, across species. They report that "mean ENC across all sampled nemtaode species is 46.7 +/- 5.1" but that *S. stercoralis* and *S. ratti* are outliers with low ENC values (~35). Note: ENC values range from 20 for highly biased to 61 for no bias. Thus, benchmarking from Mitreva *et al* 2006, we might expect that *Strongyloides* species will have greater codon bias than *C. elegans*. Unfortunately, although they do report measuring distribution of ENC values across all transcripts for each species, they only have plots for *S. ratti*, *P. trichosuri* and *P. pacificus*.   
```{r caiDist}

species.specific.dat.df <- dat.genomic.df %>%
  dplyr::mutate(species_CAI = case_when(
    species == 'Ce' ~ Ce_CAI,
    species != 'Ce' ~ Sr_CAI
  )) %>%
  dplyr:: select(geneID, species_CAI, species)
  

species_adaptiveness_plot <- species.specific.dat.df %>%
  ggplot(aes(x = species_CAI, y = after_stat(ndensity), color = species)) +
  geom_freqpoly(bins = 50) +
  scale_color_manual(values = c("seagreen4", "steelblue4", 
                                "coral4", "darkgoldenrod4", 
                                "darkorchid4"),
                     name = "Species",
                     labels = c("C. elegans", "S. papillosus", 
                                "S. ratti", "S. stercoralis",
                                "S. venezuelensis")) +
  labs(title = "Species-wide codon adaptiveness",
       subtitle = "Data includes all predicted coding sequences",
       x = "Codon bias relative to \n genus coding usage rules (CAI)",
       y = "Density (scaled to maximum of 1)",
       caption = "Strongyloides species values relative to S. ratti usage rules
       C. elegans values relative to C. elegans usage rules.") +
  coord_equal(xlim = c(0,1), ylim = c(0,1)) +
  theme_bw() +
  theme(plot.title.position = "plot",
        plot.caption.position = "panel",
        plot.title = element_text(face = "bold",
                                  size = 12, hjust = 0),
        plot.subtitle = element_text(size = 10),
        legend.title = element_text(size = 10),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(angle = 45, hjust = 1))

ggsave('CodonAdaptivenessDistributions.pdf', plot = species_adaptiveness_plot, 
       width = 5.5, height = 4, 
       units = "in", device = "pdf", 
       useDingbats = FALSE)

species_adaptiveness_plot
```
  
Notes re: the above plot - *C. elegans* shows overall lower codon bias compared to the *Strongyloides* species, as predicted from the Mitreva *et al* 2006 results. Reminder that Mitreva *et al* reported that this difference is greatly affected by the GC content; more AT rich speices like the *Strongyloides* species have greater codon usage biases. Also, it looks like the *Strongyloides* subclades (*S. ratti* - *S. stercoralis*; *S. papillosus* - *S. venezuelensis*) cluster together.  

## Load Genes-of-Interest Data  
For gene set of interest, load data listing GC content and CAI indeces.  
```{r loadGoIData, message=FALSE}
temp <- c(Ss_GoI = './Ss_Chemoreceptor_Codon_Usage_Report.xlsx',
          Sr_GoI = './Sr_Chemoreceptor_Codon_Usage_Report.xlsx',
          Sv_GoI = './Sv_Chemoreceptor_Codon_Usage_Report.xlsx',
          Ce_GoI = './Ce_Chemoreceptor_Codon_Usage_Report.xlsx')

dat.Ss.GoI <- read_excel(temp[["Ss_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Ss")

dat.Sr.GoI <- read_excel(temp[["Sr_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Sr")

dat.Sv.GoI <- read_excel(temp[["Sv_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Sv")

dat.Ce.GoI <- read_excel(temp[["Ce_GoI"]], 
                         sheet = 1, 
                         skip = 3, 
                         col_names = T) %>%
  dplyr::select(!c(GC,`cDNA sequence`)) %>%
  dplyr::mutate(species = "Ce")

dat.GoI.df <- full_join(dat.Ss.GoI, dat.Sr.GoI, by = NULL) %>%
  full_join(dat.Sv.GoI, by = NULL) %>%
  full_join(dat.Ce.GoI, by = NULL)

```

## Plot Genes-of-Interest CAI values relative to Genomic data  
Generate a series of plots (separated by species ID) where for each gene the codon adaptation index relative to highly expressed *S. ratti* transcripts is plotted against the codon adaptation index relative to highly expressed *C. elegans* genes. For every plot, also include the values for the genes-of-interest set, so we can compare how the values of those genes fall compared to all the other genes in that respective species' genome.  
```{r plotGoIData}

tbl <- dat.genomic.df

cai_plot <- ggplot(tbl, aes(Sr_CAI, Ce_CAI, species)) +
  geom_smooth(method=lm, color = "steelblue4", 
              fill = "steelblue1", linetype = 2, 
              size = 0.5, formula = "y ~ x") +
  geom_point(shape = 1, size = 2, alpha = 0.5) +
  geom_point(dat.GoI.df, mapping = aes(Sr_CAI, Ce_CAI, color = species),
             show.legend = F,
             shape = 1, size = 2, alpha = 1) +
  scale_color_manual(values = c("seagreen4", "coral4", "darkgoldenrod4", "darkorchid4"))+
  
  geom_hline(yintercept = 0.5, color = "grey", linetype = 2) +
  geom_vline(xintercept = 0.5, color = "grey", linetype = 2) +
  # geom_point(dplyr::select(dat.Ce.GoI,!species), mapping = aes(Sr_CAI, Ce_CAI),
  #             show.legend = T,
  #             shape = 1, size = 2, alpha = 1, color = "seagreen4") +
  # geom_point(dplyr::select(dat.Ss.GoI,!species), mapping = aes(Sr_CAI, Ce_CAI),
  #             show.legend = T,
  #             shape = 1, size = 2, alpha = 1, color = "coral4") +
  facet_grid(~species) +
  labs(title = "Species-specific codon adaptiveness",
       subtitle = "colored icons = species-specific chemoreceptors; black icons = all coding sequences
         ",
       x = "Codon bias relative to \n S. ratti usage rules (CAI)",
       y = "Codon Bias relative to \n C. elegans usage rules (CAI)",
       caption = "Blue line/shading = linear regression \n w/ 95% confidence regions; \n formula = y ~ x") +
  coord_equal(xlim = c(0,1), ylim = c(0,1)) +
  theme_bw() +
  theme(plot.title.position = "plot",
        plot.caption.position = "panel",
        plot.title = element_text(face = "bold",
                                  size = 12, hjust = 0),
        plot.subtitle = element_text(size = 10),
        axis.title = element_text(size = 10),
        axis.text = element_text(size = 10),
        axis.text.x = element_text(angle = 45, hjust = 1))

ggsave('ChemosensoryCodonUsagePlot.pdf', plot = cai_plot, 
       width = 11, height = 8, 
       units = "in", device = "pdf", 
       useDingbats = FALSE)

cai_plot

```

# Appendix: All code for this report  
```{r ref.label=knitr::all_labels(), echo=TRUE, eval=FALSE}
```
